test/unit/math/rev/mat/functor/algebra_solver_fp_test.cpp

```diff
@@ -0,0 +1,594 @@
+          fp;  // NOLINT
+      double f_tol = 1.e-12;
+      int max_num_steps = 100;
+      return fp.solve(x, y_, env, f_tol, max_num_steps)(i);
+    };
+    return f_fd;
+  }
+};
+
+TEST_F(FP_exp_func_test, solve) {
+  KinsolFixedPointEnv<FP_exp_func> env(f, x, y, x_r, x_i, msgs, u_scale,
+                                       f_scale);
+  FixedPointSolver<KinsolFixedPointEnv<FP_exp_func>, FixedPointADJac> fp;
+  double f_tol = 1.e-12;
+  int max_num_steps = 100;
+  {
+    Eigen::Matrix<double, -1, 1> res
+        = fp.solve(x, y, env, f_tol, max_num_steps);  // NOLINT
+    EXPECT_FLOAT_EQ(res(0), 0.567143290409);
+  }
+
+  {
+    x(0) = 0.1;
+    y(0) = 0.8;
+    Eigen::Matrix<double, -1, 1> res
+        = fp.solve(x, y, env, f_tol, max_num_steps);  // NOLINT
+    EXPECT_FLOAT_EQ(res(0), 0.612584823501);
+  }
+}
+
+TEST_F(FP_2d_func_test, solve) {
+  KinsolFixedPointEnv<FP_2d_func> env(f, x, y, x_r, x_i, msgs, u_scale,
+                                      f_scale);
+  FixedPointSolver<KinsolFixedPointEnv<FP_2d_func>, FixedPointADJac> fp;
+  double f_tol = 1.e-12;
+  int max_num_steps = 100;
+  Eigen::Matrix<double, -1, 1> res
+      = fp.solve(x, y, env, f_tol, max_num_steps);  // NOLINT
+  EXPECT_NEAR(res(0), 0.7861518, 1e-5);
+  EXPECT_NEAR(res(1), 0.6180333, 1e-5);
+}
+
+TEST_F(FP_exp_func_test, gradient) {
+  KinsolFixedPointEnv<FP_exp_func> env(f, x, y, x_r, x_i, msgs, u_scale,
+                                       f_scale);
+  FixedPointSolver<KinsolFixedPointEnv<FP_exp_func>, FixedPointADJac> fp;
+  double f_tol = 1.e-12;
+  int max_num_steps = 100;
+  Eigen::Matrix<var, -1, 1> yp(to_var(y));
+
+  x(0) = 0.1;
+  y(0) = 0.8;
+  Eigen::Matrix<var, -1, 1> x_sol
+      = fp.solve(x, yp, env, f_tol, max_num_steps);  // NOLINT
+  EXPECT_FLOAT_EQ(value_of(x_sol(0)), 0.612584823501);
+  stan::math::set_zero_all_adjoints();
+  x_sol(0).grad();
+  auto f_fd = fd_functor(0);
+  double fx;
+  Eigen::VectorXd grad_fx;
+  finite_diff_gradient_auto(f_fd, y, fx, grad_fx);
+  EXPECT_FLOAT_EQ(grad_fx(0), yp(0).adj());
+}
+
+TEST_F(FP_2d_func_test, gradient) {
+  KinsolFixedPointEnv<FP_2d_func> env(f, x, y, x_r, x_i, msgs, u_scale,
+                                      f_scale);
+  FixedPointSolver<KinsolFixedPointEnv<FP_2d_func>, FixedPointADJac> fp;
+  double f_tol = 1.e-12;
+  int max_num_steps = 100;
+  Eigen::Matrix<var, -1, 1> yp(to_var(y));
+
+  Eigen::Matrix<var, -1, 1> x_sol
+      = fp.solve(x, yp, env, f_tol, max_num_steps);  // NOLINT
+  EXPECT_NEAR(value_of(x_sol(0)), 0.7861518, 1e-5);
+  EXPECT_NEAR(value_of(x_sol(1)), 0.6180333, 1e-5);
+
+  double fx;
+  Eigen::VectorXd grad_fx;
+  for (int i = 0; i < env.N_; ++i) {
+    stan::math::set_zero_all_adjoints();
+    x_sol(i).grad();
+    auto f_fd = fd_functor(i);
+    finite_diff_gradient_auto(f_fd, y, fx, grad_fx);
+    for (int j = 0; j < env.M_; ++j) {
+      EXPECT_FLOAT_EQ(grad_fx(j), yp(j).adj());
+    }
+  }
+}
+
+TEST_F(FP_2d_func_test, gradient_with_var_init_point) {
```

Comment:
Having a var initial point (initial guess) shouldn't affect the gradient.

Reply:
True. But we still want to test it otherwise the branch of template specialization won't be covered by tests.

Reply:
Ok, that makes sense. The name of the test function is then misleading. Add a comment clarifying the purpose of the test and we're good.

Reply:
I've edited the documentation for initial type in `algebra_solver_fp` signature instead:
```
 * @tparam T type of initial guess vector. The final soluton
 *           type doesn't depend on initial guess type,
 *           but we allow initial guess to be either data or param.
```